nix-channels.py

```diff
@@ -0,0 +1,298 @@
+
+        if tagline_res is None:
+            logging.warning(f'    - Invalid tagline: {tagline}')
+            continue
+
+        released_time = tagline_res[1]
+
+        release_path.mkdir(parents=True, exist_ok=True)
+
+        with (release_path / '.released-time').open('w') as f:
+            f.write(released_time)
+
+        logging.info(f'    - Downloading files')
+
+        has_hash_fail = False
+
+        for row in node('tr'):
+            td = pq(row)('td')
+            if len(td) != 3:
+                continue
+            file_name, _file_size, file_hash = (pq(x).text() for x in td)
+
+            if file_name.endswith('.ova') or file_name.endswith('.iso'):
+                # Skip images
+                pass
+            elif (release_path / file_name).exists() \
+                and file_sha256(release_path / file_name) == file_hash:
+                logging.info(f'      - {file_name} (existing)')
+            else:
+                if file_name == 'binary-cache-url':
+                    logging.info(f'      - binary-cache-url (redirected)')
+                    dest = '.original-binary-cache-url'
+                else:
+                    logging.info(f'      - {file_name}')
+                    dest = file_name
+
+                download(f'{chan_location}/{file_name}', release_path / dest)
+                if file_sha256(release_path / dest) != file_hash:
+                    global failure
+                    failure = True
+
+                    has_hash_fail = True
+                    logging.error(f'        Wrong hash!')
+                    logging.error(f'        - expected {file_hash}')
+                    logging.error(f'        - got      {hash}')
+
+        logging.info('    - Writing binary-cache-url')
+        (release_path / 'binary-cache-url').write_text(binary_cache_url)
+
+        if has_hash_fail:
+            logging.warning('    - Found bad files. Not updating symlink.')
+        else:
+            channels_to_update.append(channel)
+            if chan_path_update.exists():
+                chan_path_update.unlink()
+            chan_path_update.symlink_to(release_target)
+
+            logging.info(f'    - Symlink updated')
+
+    return channels_to_update
+
+def update_channels(channels):
+    logging.info(f'- Updating binary cache')
+
+    for channel in channels:
+        logging.info(f'  - {channel}')
+
+        chan_path_update = working_dir / f'.{channel}.update'
+
+        upstream_binary_cache = (chan_path_update / '.original-binary-cache-url').read_text()
+
+        with lzma.open(str(chan_path_update / 'store-paths.xz')) as f:
+            paths = [ path.rstrip() for path in f ]
+
+        logging.info(f'    - {len(paths)} paths listed')
+
+        # xargs can splits up the argument lists and invokes nix copy multiple
+        # times to avoid E2BIG (Argument list too long)
+        nix_process = subprocess.Popen(
+            [ 'xargs', 'nix', 'copy',
```

Comment:
What `nix copy` does? Can we avoid some external commands. Otherwise you need to install it on the mirror server, or in the base docker image.

Reply:
`nix copy` is necessary for operate on a Nix store. I think docker image is a good approach.

NixOS provide a way to generate docker image via a nix expression, so a we can create a nix package for the sync script and create the docker image easily based on this package.

Reply:
`nix copy` in this context is essentially the official way to mirror. (Its more general use is copying store paths from one store to another, and in this case we are copying all paths in a channel from upstream to a local binary cache.) If we implement this ourselves it will need to find dependencies recursively by parsing the metadata and skip already existing paths, so there's quite a bit of logic to duplicate.

I believe (but have not tried) that installing Nix in an existing docker image is as easy as copying Nix itself in (Nix and deps from `/nix/store`). Nix wouldn't be a functioning package manager, but `nix copy` should work.